quantum/process_keycode/process_unicode_common.c

```diff
@@ -19,42 +19,64 @@
 #include <string.h>
 #include <ctype.h>
 
-static uint8_t input_mode;
-uint8_t mods;
+unicode_config_t unicode_config;
+#if UNICODE_SELECTED_MODES != -1
+static uint8_t selected[] = { UNICODE_SELECTED_MODES };
+static uint8_t selected_count = sizeof selected / sizeof *selected;
+static uint8_t selected_index;
+#endif
+static uint8_t saved_mods;
 
 void set_unicode_input_mode(uint8_t os_target) {
-  input_mode = os_target;
-  eeprom_update_byte(EECONFIG_UNICODEMODE, os_target);
+  unicode_config.input_mode = os_target;
+  persist_unicode_input_mode();
+  dprintf("input_mode set to: %u\n", unicode_config.input_mode);
 }
 
 uint8_t get_unicode_input_mode(void) {
-  return input_mode;
+  return unicode_config.input_mode;
 }
 
 void unicode_input_mode_init(void) {
-  static bool first_flag = false;
-  if (!first_flag) {
-    input_mode = eeprom_read_byte(EECONFIG_UNICODEMODE);
-    first_flag = true;
+  unicode_config.raw = eeprom_read_byte(EECONFIG_UNICODEMODE);
+#if UNICODE_SELECTED_MODES != -1
+  // Find input_mode in selected modes
+  uint8_t i;
+  for (i = 0; i < selected_count; i++) {
+    if (selected[i] == unicode_config.input_mode) {
+      selected_index = i;
+      break;
+    }
+  }
+  if (i == selected_count) {
+    // input_mode isn't selected, change to one that is
+    unicode_config.input_mode = selected[selected_index = 0];
   }
+#endif
+  dprintf("input_mode init to: %u\n", unicode_config.input_mode);
```

Comment:
@drashna @mechmerlin Is it okay to have these debug prints in core code, or should I remove / comment them out?